File: week12/19/P4.py
class LinkedNode():
    def __init__(self, x):
        self.val = x
        self.next = None

class MyQueue(object):

    def __init__(self):
        self.sentinel = LinkedNode(-999)
        self.size = 0    

    def push(self, x):
        curNode = self.sentinel
        while curNode.next:
            curNode = curNode.next
        curNode.next = LinkedNode(x)
        self.size += 1

    def pop(self):
        if self.empty():
            return None
        front = self.sentinel.next.val
        self.sentinel.next = self.sentinel.next.next
        self.size -= 1
        return front

    def peek(self):
        if self.empty():
            return None
        return self.sentinel.next.val

    def empty(self):
        return self.size == 0
        
if __name__ == "__main__":
    obj = MyQueue()
    obj.push(1)
    obj.push(2)
    print(obj.peek())
    print(obj.pop())
    print(obj.empty())

File: week12/19/P4.cpp
#include <iostream>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        this->next = nullptr;
    }
};

class MyQueue {
public:
    Node* head;

    MyQueue() {
        head = nullptr;
    }

    void push(int x) {
        Node* new_node = new Node(x);
        if (head == nullptr) head = new_node; return;
        Node* last_node = head;
        while (last_node->next != nullptr) {
            last_node = last_node->next;
        }
        last_node->next = new_node;
    }
    
    int pop() {
        if (head == nullptr) return -1;
        int front = head->data;
        head = head->next;
        return front;
    }
    
    int peek() {
        if (head == nullptr) return -1;
        return head->data;
    }
    
    bool empty() {
        return head == nullptr;
    }
};

int main() {
    MyQueue* obj = new MyQueue();
    obj->push(1);
    obj->push(2);
    cout << obj->peek() << endl;
    cout << obj->pop() << endl;
    cout << obj->empty() << endl;

    return 0;
}